packages/x-charts/src/ChartContainer/useDefaultizeAxis.ts

```diff
@@ -16,15 +18,28 @@ const defaultizeAxis = (
     ...(inAxis === undefined || inAxis.findIndex(({ id }) => id === DEFAULT_AXIS_KEY) === -1
       ? [{ id: DEFAULT_AXIS_KEY, scaleType: 'linear' as const }]
       : []),
-  ];
+  ].map((axisConfig) => {
+    const dataKey = axisConfig.dataKey;
+    if (dataKey === undefined || axisConfig.data !== undefined) {
+      return axisConfig;
+    }
+    if (dataset === undefined) {
+      throw Error(`MUI X: ${axisName}-axis uses \`dataKey\` but no \`dataset\` is provided.`);
+    }
+    return {
+      ...axisConfig,
+      data: dataset.map((d) => d[dataKey]),
+    };
+  });
 };
 
 export const useDefaultizeAxis = (
   inXAxis: MakeOptional<AxisConfig<ScaleName, any, ChartsAxisProps>, 'id'>[] | undefined,
   inYAxis: MakeOptional<AxisConfig<ScaleName, any, ChartsAxisProps>, 'id'>[] | undefined,
+  dataset: DatasetType | undefined,
 ) => {
-  const xAxis = React.useMemo(() => defaultizeAxis(inXAxis, 'x'), [inXAxis]);
-  const yAxis = React.useMemo(() => defaultizeAxis(inYAxis, 'y'), [inYAxis]);
+  const xAxis = React.useMemo(() => defaultizeAxis(inXAxis, dataset, 'x'), [inXAxis, dataset]);
```

Comment:
For me this would be clearer but maybe it causes TS issues. DOes not really matter

```suggestion
const xAxis = React.useMemo(() => normalize(defaultizeAxis(inXAxis, 'x'), dataset, 'x'), [inXAxis, dataset]);
```